src/utils.py

```diff
@@ -429,6 +429,109 @@ def powerset(seq: Sequence, exclude_empty: bool) -> Iterator[Sequence]:
                                          for r in range(start, len(seq)+1))
 
 
+_S = TypeVar('_S', bound=Hashable)  # state in heuristic search
+_A = TypeVar('_A')  # action in heuristic search
+
+
+@dataclass(frozen=True)
+class _HeuristicSearchNode(Generic[_S, _A]):
+    state: _S
+    edge_cost: float
+    cumulative_cost: float
+    parent: Optional[_HeuristicSearchNode[_S, _A]] = None
+    action: Optional[_A] = None
+
+
+def _run_heuristic_search(
+    initial_state: _S,
+    check_goal: Callable[[_S], bool],
+    get_successors: Callable[[_S], Iterator[Tuple[_A, _S, float]]],
+    get_priority: Callable[[_HeuristicSearchNode[_S, _A]], Any],
+    max_expansions: int = 1000
+    ) -> Tuple[List[_S], List[_A]]:
+    """A generic heuristic search implementation.
+
+    Depending on get_priority, can implement A*, GBFS, or UCS.
+
+    If no goal is found, returns the node with the best priority.
```

Comment:
Could you update this comment to describe what's actually returned? It's not the node, since the node gets passed through finish_plan first.